storagemarket/impl/shardregistration.go

```diff
@@ -0,0 +1,146 @@
+	"github.com/filecoin-project/specs-storage/storage"
+
+	mktdagstore "github.com/filecoin-project/go-fil-markets/dagstore"
+	"github.com/filecoin-project/go-fil-markets/storagemarket"
+	"github.com/filecoin-project/go-fil-markets/storagemarket/impl/providerstates"
+)
+
+var shardRegKey = datastore.NewKey("shards-registered")
+
+type SectorState interface {
+	StateSectorGetInfo(context.Context, address.Address, abi.SectorNumber, types.TipSetKey) (*miner.SectorOnChainInfo, error)
+	IsUnsealed(ctx context.Context, sector storage.SectorRef, offset storiface.UnpaddedByteIndex, size abi.UnpaddedPieceSize) (bool, error)
+}
+
+type ShardRegistration struct {
+	maddr       address.Address
+	ds          datastore.Datastore
+	dagStore    mktdagstore.DagStoreWrapper
+	sectorState SectorState
+}
+
+func NewShardRegistration(
+	maddr address.Address,
+	ds datastore.Datastore,
+	dagStore mktdagstore.DagStoreWrapper,
+	sectorState SectorState,
+) *ShardRegistration {
+	return &ShardRegistration{
+		maddr:       maddr,
+		ds:          ds,
+		dagStore:    dagStore,
+		sectorState: sectorState,
+	}
+}
+
+func (r *ShardRegistration) registerShards(ctx context.Context, deals []storagemarket.MinerDeal) error {
+	// Check if all deals have already been registered as shards
+	has, err := r.ds.Has(shardRegKey)
+	if err != nil {
+		return xerrors.Errorf("failed to get shard registration status: %w", err)
+	}
+	if has {
+		// All deals have been registered as shards, bail out
+		return nil
+	}
+
+	// Filter for deals that are currently sealing.
+	// If the deal has not yet been handed off to the sealing subsystem, we
+	// don't need to call RegisterShard in this migration; RegisterShard will
+	// be called in the new code once the deal reaches the state where it's
+	// handed off to the sealing subsystem.
+	resch := make(chan dagstore.ShardResult, len(deals))
+	for _, deal := range deals {
+		if deal.Ref.PieceCid == nil {
+			continue
+		}
+
+		// Check if the deal has been handed off to the sealing subsystem
+		var sealing bool
+		for _, state := range providerstates.ProviderSealingStates {
+			if deal.State == state {
+				sealing = true
+				break
+			}
+		}
+		if !sealing {
+			continue
+		}
+
+		// Check if the deal is in an unsealed state
+		isUnsealed, err := r.isUnsealed(ctx, deal.SectorNumber)
+		if err != nil {
+			log.Errorf("failed to get unsealed state of deal with piece CID %s: %s", deal.Ref.PieceCid, err)
+		}
+
+		// Register the deal as a shard with the DAG store, initializing the
+		// index immediately if the deal is unsealed (if the deal is not
+		// unsealed it will be initialized "lazily" once it's unsealed during
+		// retrieval)
+		r.dagStore.RegisterShardAsync(ctx, *deal.Ref.PieceCid, deal.CARv2FilePath, isUnsealed, resch)
+	}
+
+	// If there are any problems registering shards, just log an error
+	go func() {
+		for res := range resch {
+			if res.Error != nil {
+				log.Errorf("failed to register shard: %s", res.Error)
+			}
+		}
```

Comment:
Better to start this goroutine should be started _before_ we perform the loop. Then we can get rid of the buffered channel and use an unbuffered one. (A buffered channel when there are, say 1MM deals, can get costly).

Reply:
I think somewhere we're going to have to keep a buffer:
- we could create a copy of the deals array and then filter in place, but that requires as much space as the original array
- we could create a new slice and append to it each time we register a deal but that is going to be much slower than reserving the buffer in advance
- we could run over the loop twice, which would save memory but at the cost of CPU

I don't think anyone has made a million deals, but even if they had the buffered channel would only require 1m pointers worth of space - so I think the buffered channel may be the least-bad solution here.

Reply:
If you start the goroutine before you start registering deals, you can make progress on the results consumption in parallel, and you can remove the channel buffer entirely, or keep it small. Right now you're using the channel buffer as a slice, and that's mildly inefficient.

Reply:
The problem is that in order to know when to stop reading from the channel, you need to know the total number of deals. And you don't know that until the end of the for loop. So if you try to make progress on results consumption in parallel you're in a race between the go routine and the for loop.

Reply:
Pushed a solution in https://github.com/filecoin-project/go-fil-markets/commit/f391a57f3af3dd30e79a1052dfcbf1924cb52be6.

Alternatively we could’ve done one iteration where we decided which shards to register (and thus get the size), another to actually register them.